docs/source/cpp/csv.rst

```diff
@@ -0,0 +1,142 @@
+      arrow::MemoryPool* pool = default_memory_pool();
+      std::shared_ptr<arrow::io::InputStream> input = ...;
+
+      auto read_options = arrow::csv::ReadOptions::Defaults();
+      auto parse_options = arrow::csv::ParseOptions::Defaults();
+      auto convert_options = arrow::csv::ConvertOptions::Defaults();
+
+      // Instantiate TableReader from input stream and options
+      std::shared_ptr<arrow::csv::TableReader> reader;
+      st = arrow::csv::TableReader::Make(pool, input, read_options,
+                                         parse_options, convert_options,
+                                         &reader);
+      if (!st.ok()) {
+         // Handle TableReader instantiation error...
+      }
+
+      std::shared_ptr<arrow::Table> table;
+      // Read table from CSV file
+      st = reader->Read(&table);
+      if (!st.ok()) {
+         // Handle CSV read error
+         // (for example a CSV syntax error or failed type conversion)
+      }
+   }
+
+Column names
+============
+
+There are three possible ways to infer column names from the CSV file:
+
+* By default, the column names are read from the first row in the CSV file
+* If :member:`ReadOptions::column_names` is set, it forces the column
+  names in the table to these values (the first row in the CSV file is
+  read as data)
+* If :member:`ReadOptions::autogenerate_column_names` is true, column names
+  will be autogenerated with the pattern "f0", "f1"... (the first row in the
+  CSV file is read as data)
+
+Column selection
+================
+
+By default, Arrow reads all columns in the CSV file.  You can narrow the
+selection of columns with the :member:`ConvertOptions::include_columns`
+option.  If some columns in :member:`ConvertOptions::include_columns`
+are missing from the CSV file, an error will be emitted unless
+:member:`ConvertOptions::include_missing_columns` is true, in which case
+the missing columns are assumed to contain all-null values.
+
+Interaction with column names
+-----------------------------
+
+If both :member:`ReadOptions::column_names` and :member:`ConvertOptions::include_columns`,
+the :member:`ReadOptions::column_names` are assumed to map to CSV columns,
+and :member:`ConvertOptions::include_columns` is a subset of those column
+names that will part of the Arrow Table.
+
+Data types
+==========
+
+It is possible to convert a CSV column into the following Arrow data types:
+
+* Null
+* All Integer types
+* Float32 and Float64
+* Decimal128
+* Boolean
+* Timestamp
+* Binary and Large Binary
+* String and Large String (with optional UTF8 input validation)
+* Fixed-Size Binary
+
+By default, the CSV reader infers the most appropriate data type for each
+column.  Type inference considers the following data types, in order:
+
+* Null
+* Int64
+* Boolean
+* Timestamp (with seconds unit)
+* Float64
+* String
+* Binary
+
+It is possible to override type inference for select columns by setting
+the :member:`ConvertOptions::column_types` option.
```

Comment:
So there are some types I still can't get from the CSV reader even if I specify `column_types`?

Reply:
Yes, you can't for instance convert a CSV value to an Arrow struct or list or map... Is that what you were asking?

Reply:
Yes, IMO it would be good to make that explicit, that the limitation on types is not just about inferring them from the data, there are some types you can't even specify.

Reply:
I can add a sentence that non-listed types will error out.